sweep/sweeper.go

```diff
@@ -30,13 +73,585 @@ type UtxoSweeperConfig struct {
+// txes.
+func (s *UtxoSweeper) Stop() error {
+	if !atomic.CompareAndSwapUint32(&s.stopped, 0, 1) {
+		return nil
+	}
+
+	log.Infof("Sweeper shutting down")
+
+	close(s.quit)
+	s.wg.Wait()
+
+	log.Infof("Sweeper shut down")
+
+	return nil
+}
+
+// SweepInput sweeps inputs back into the wallet. The inputs will be batched and
+// swept after the batch time window ends.
+func (s *UtxoSweeper) SweepInput(input Input) (chan Result, error) {
+	log.Debugf("Sweep request: %v", input.OutPoint())
+
+	sweeperInput := &sweeperInput{
+		input:      input,
+		resultChan: make(chan Result, 1),
+	}
+
+	// Deliver input to main event loop.
+	select {
+	case s.newInputs <- sweeperInput:
+	case <-s.quit:
+		return nil, fmt.Errorf("sweeper shutting down")
+	}
+
+	return sweeperInput.resultChan, nil
+}
+
+// collector is the sweeper main loop. It processes new inputs, spend
+// notifications and counts down to publication of the sweep tx.
+func (s *UtxoSweeper) collector() {
+	defer s.wg.Done()
+	defer s.cfg.Ticker.Stop()
+
+	log.Debug("Collector started")
+
+	for {
+		select {
+		case input := <-s.newInputs:
+			outpoint := *input.input.OutPoint()
+			pendInput, pending := s.pendingInputs[outpoint]
+			if pending {
+				log.Debugf("Already pending input %v received",
+					outpoint)
+
+				// Add additional result channel to signal spend
+				// of this input.
+				pendInput.listeners =
+					append(pendInput.listeners,
+						input.resultChan)
+
+				break
+			}
+
+			// Create a new pendingInput and initialize the
+			// listeners slice with the passed in result channel. If
+			// this input is offered for sweep again, the result
+			// channel will be appended to this slice.
+			pendInput = &pendingInput{
+				listeners: []chan Result{input.resultChan},
+				input:     input.input,
+				state:     stateNew,
+			}
+			s.pendingInputs[outpoint] = pendInput
+
+			// Precalculate weight estimate, as it may be used
+			// needed multiple times. Also use it to signal the
+			// listener early in case of an unexpected witness type.
+			var weightEstimate lnwallet.TxWeightEstimator
+			err := s.addInputWeight(&weightEstimate, input.input)
+			if err != nil {
+				s.signalAndRemove(&outpoint, Result{Err: fmt.Errorf(
+					"cannot get weight estimate: %v",
+					err)})
+
+				break
+			}
+			pendInput.addedWeight = weightEstimate.Weight()
+
+			// Start watching for spend of this input, either by us
+			// or the remote party. The possibility of a remote
+			// spend is there reason why just registering for conf.
+			// of our sweep tx isn't enough.
+			pendInput.ntfnRegCancel, err = s.waitForSpend(input.input)
+			if err != nil {
+				s.signalAndRemove(&outpoint, Result{Err: fmt.Errorf(
+					"cannot watch for spend: %v",
+					err)})
+
+				break
+			}
+
+			// Lookup if there is already a confirmed or unconfirmed
+			// tx of ours that spends this input.
+			spendingTx, err :=
+				s.cfg.HasSpendingTx(outpoint)
```

Comment:
So the reason we need `HasSpendingTx` instead of `RegisterForSpend` is that we need to distinguish whether we are the spender?

Reply:
`HasSpendingTx` is only needed after restarts. We use the wallet to find out if we are already sweeping an input, so that we don't publish another tx for it (and use up another address).